File: backend/danger.py
"""Danger Loop — regex risk engine. SAFE < LOW < MEDIUM < HIGH < CRITICAL."""
import re
from typing import Literal

RiskLevel = Literal["safe", "low", "medium", "high", "critical"]

# ponytail: pure regex blocklist, not a sandbox. Upgrade path: allowlist + seccomp/job objects.
_PATTERNS: dict[str, list[str]] = {
    "critical": [
        r"rm\s+-rf\s+/", r"rm\s+-rf\s+~", r"rmdir\s+/s\s+/q",
        r"format\s+[cCdD]:", r"del\s+/[sS]\s+[cC]:\\",
        r"mkfs\.", r"dd\s+if=.*of=/dev/",
    ],
    "high": [
        r"rm\s+-rf\s+", r"rm\s+-r\s+", r"rmdir\s+", r"del\s+/[sS]\s+",
        r"sudo\s+rm", r"chmod\s+777", r"kill\s+-9", r"taskkill\s+/f",
        r"remove-item\s+-recurse", r"remove-item\s+.*-force",
    ],
    "medium": [
        r"\brm\s+", r"\bdel\s+", r"\bmv\s+", r"\bchmod\s+", r"\bchown\s+",
        r"\bsudo\b", r"net\s+user", r"net\s+localgroup",
    ],
    "low": [
        r"mkdir\s+", r"touch\s+", r"echo\s+.*>", r"\bcp\s+", r"\bcopy\s+",
    ],
}

_COMPILED = {lvl: [re.compile(p, re.IGNORECASE) for p in pats] for lvl, pats in _PATTERNS.items()}
_ORDER = ["critical", "high", "medium", "low"]


def assess(command: str) -> RiskLevel:
    cmd = (command or "").strip()
    for level in _ORDER:
        if any(rx.search(cmd) for rx in _COMPILED[level]):
            return level  # type: ignore[return-value]
    return "safe"


if __name__ == "__main__":
    cases = {
        "rm -rf /": "critical",
        "mkfs.ext4 /dev/sda1": "critical",
        "Remove-Item -Recurse -Force C:\\Users": "high",
        "format C:": "critical",
        "rm -rf build/": "high",
        "taskkill /f /IM app.exe": "high",
        "rm notes.txt": "medium",
        "sudo apt install": "medium",
        "mkdir test": "low",
        "echo hi > a.txt": "low",
        "ls -la": "safe",
        "git status": "safe",
    }
    for cmd, expected in cases.items():
        got = assess(cmd)
        assert got == expected, f"{cmd!r}: expected {expected}, got {got}"
    print(f"danger.py: {len(cases)} self-checks passed")

File: src-tauri/src/danger.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum RiskLevel {
    Safe,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandResult {
    pub stdout: String,
    pub stderr: String,
    pub requires_confirmation: bool,
    pub risk_level: String,
}

pub fn assess_risk(cmd: &str) -> RiskLevel {
    let lower = cmd.trim().to_lowercase();

    // Critical patterns
    if lower.contains("rm -rf /") || lower.contains("rm -rf ~") || lower.contains("rmdir /s /q")
        || lower.contains("format c:") || lower.contains("format d:")
        || lower.contains("mkfs.") || lower.contains("dd if=")
    {
        return RiskLevel::Critical;
    }

    // High patterns
    if lower.contains("rm -rf ") || lower.contains("rm -r ") || lower.contains("sudo rm")
        || lower.contains("chmod 777") || lower.contains("kill -9")
        || lower.contains("taskkill /f") || lower.contains("remove-item -recurse")
        || lower.contains("remove-item") && lower.contains("-force")
    {
        return RiskLevel::High;
    }

    // Medium patterns
    if lower.contains("sudo ") || lower.contains("rm ") || lower.contains("del ")
        || lower.contains("mv ") || lower.contains("chmod ") || lower.contains("chown ")
        || lower.contains("net user") || lower.contains("net localgroup")
    {
        return RiskLevel::Medium;
    }

    // Low patterns
    if lower.contains("mkdir ") || lower.contains("touch ") || lower.contains("echo ")
        || lower.contains("cp ") || lower.contains("copy ")
    {
        return RiskLevel::Low;
    }

    RiskLevel::Safe
}

pub fn execute_command(cmd: &str) -> CommandResult {
    let risk = assess_risk(cmd);

    if risk >= RiskLevel::Critical {
        return CommandResult {
            stdout: String::new(),
            stderr: "Command blocked: too dangerous.".to_string(),
            requires_confirmation: false,
            risk_level: format!("{:?}", risk),
        };
    }

    if risk >= RiskLevel::Medium {
        return CommandResult {
            stdout: String::new(),
            stderr: String::new(),
            requires_confirmation: true,
            risk_level: format!("{:?}", risk),
        };
    }

    let output = std::process::Command::new("cmd")
        .args(["/C", cmd])
        .output();

    match output {
        Ok(out) => CommandResult {
            stdout: String::from_utf8_lossy(&out.stdout).to_string(),
            stderr: String::from_utf8_lossy(&out.stderr).to_string(),
            requires_confirmation: false,
            risk_level: format!("{:?}", risk),
        },
        Err(e) => CommandResult {
            stdout: String::new(),
            stderr: format!("Execution error: {}", e),
            requires_confirmation: false,
            risk_level: format!("{:?}", risk),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_risk_levels() {
        assert_eq!(assess_risk("ls"), RiskLevel::Safe);
        assert_eq!(assess_risk("mkdir test"), RiskLevel::Low);
        assert_eq!(assess_risk("rm file.txt"), RiskLevel::Medium);
        assert_eq!(assess_risk("sudo rm -rf /"), RiskLevel::Critical);
    }

    #[test]
    fn test_execute_safe() {
        let result = execute_command("dir");
        assert!(!result.requires_confirmation);
        assert_eq!(result.risk_level, "Safe");
    }

    #[test]
    fn test_execute_critical_blocked() {
        let result = execute_command("rm -rf /");
        assert!(!result.requires_confirmation);
        assert!(result.stderr.contains("blocked"));
    }
}
